### ISSUE TITLE
Optimize Verbs With Implicit Taking

### ISSUE THREAD

**Issue by tajmone** [AnssiR66/AlanStdLib#78]
Remove useless IF clause to check that implicit-take went well in various verbs, for it's not needed.

- [x] Find all verbs that have implicit-take and an IF clause to check that it succeeded.
- [x] Remove the IF clause surrounding the code to be executed after the implicit take, and add a comment about the VERB aborting if the implicit take action fails.
- [x] Find if other verbs have similar redundant IF clauses, in contexts other than implicit taking.

### Rationale

Various verbs that carry out implicit-taking use an IF clause to check that the implicit take succeeded. E.g. from `lib_classes.i`, verb `give` on `liquid`:

```alan
  VERB give -- LIQUID
    WHEN obj
    DOES ONLY
      -- >>> implicit take >>>
      IF THIS NOT IN hero
        THEN
          IF vessel OF THIS = null_vessel OR vessel OF THIS IS NOT takeable
            THEN "You can't carry" SAY THE THIS. "around in your bare hands."
            ELSE LOCATE vessel OF THIS IN hero.
              "(taking" SAY THE vessel OF THIS. "of" SAY THIS. "first)$n"
          END IF.
      END IF.
      -- <<< implicit take <<<

      IF THIS IN hero
        -- i.e. if the implicit taking was successful
        THEN
          "You give" SAY THE vessel OF THIS. "of" SAY THIS. "to" SAY THE recipient. "."
          LOCATE vessel OF THIS IN recipient.
      END IF.
```

The problem here is that there's no need to carry out that check, for if the `LOCATE` statement in the implicit-take action were to fail for any reason (e.g. the object is held by an actor) then __the rest of the verb execution would be aborted__, so there's no risk (nor chance) that the rest of the `DOES` body is executed anyhow — therefore the whole `IF` clause surrounding the final part of the verb is unneeded.

Unfortunately, this aspect of Alan `VERB`s is not covered clearly in the _Alan Manual_, but here is a proof of this:

```alan
--==============================================================================
-- "Test Verb Execution Abortion" by Tristano Ajmone
--==============================================================================

THE room IsA LOCATION
END THE room.

THE box IsA object AT room.
  CONTAINER
    EXTRACT CHECK
      "You can't take things from Bruno's box, he'd be mad at you!"
END THE box.

THE ball IsA object IN box.
END THE ball.

SYNTAX take = take (obj)
  WHERE obj IsA object
    ELSE "You can only take objects."


ADD TO EVERY object
  VERB take
    DOES
      locate obj in Hero.
      "You take $+1." --> If the previous line fails this is not executed!
  END VERB take.
END ADD TO object.

Start at room.
```

**Comment by AnssiR66** [comment]
Hi Tristano!
ok, if the if-statement is redundant, sure we can get rid of it.
But in what way is it redundant? I can see your test example, but fail to see how that applies to the implicit taking.
If the implicit taking isn't successful, for example,

>GIVE WATER TO HORSE
You cannot carry the water in your bare hands

and then we remove "If obj in hero   --- i.e. implicit taking was succcessful"...
isn't the result:

>GIVE WATER TO HORSE

You cannot carry the water in your bare hands

You give the water to the horse.

-------------------
I didn't test this but why wouldn't it work like this?

-Anssi
________________________________

**Comment by tajmone** [comment]
Ciao @AnssiR66, I've tested that with the code example pasted in the Issue (as well as in test branches of the StdLib, locally), and yes, if a `locate` instruction fails then the verb simply quits — which means that no alternative routes are possible, i.e. one should verify that the LOCATE won't fail before attempting it if the verb needs to do something different in case of failure.

Is this correct @thoni56?

**Comment by thoni56** [comment]
Short answer: yes.

Long answer:

There are a couple of situations where the author don't have full control of what happens. Most of these are related to failing a check somewhere. One (can't actually think of any other right now...) is the `Locate` statement since that might fail for a few reasons, one is the container limits and checks.

We had (still have) a situation a few months back which stumped us all until I could figure out what was happening. During event execution a check failed and the stack was not cleared which over time led to a problem. (The actual bug is fixed, but exactly this issue has been on the back of my mind since.)

There is no way the author can protect against some of these checks and limits, e.g. the limits are not accessible programmatically so you would need to keep those as attributes as well (or instead) and do the limit check "manually"/explicitly.

Again, these features was designed when the execution of the game was "closer" to the players input. A `Locate` was thought to be more a direct action caused by the player, so the thinking was that interrupting the execution was the "right thing to do".

Also again, with more complex management of objects and what not in a library layer, these things are not necessarily true any more. (As a side note, it would be interesting to talk about things that the libraries tries to do that the Alan language does not provide, perhaps there are valuable language features to be found there.)

I have been thinking about ways to given the author control over, at least some of, these "interruptions". One thought would be to introduce a "try/catch"-like construct where you could catch the "exception" that a failed check might be viewed as.

Another option could be to allow "dry-running" a statement before actually executing it. Something like

    If Locate obj in hero Will succeeed Then ...

A variation on this might be to allow just trying or accessing the check/limit explicitly.

But it is not obvious, to me at least, if any of these is the correct solution. Any either one would be complicated to introduce in the language and in the system.

**Comment by tajmone** [comment]
Thanks for the exhaustive reply Thoni!

> We had (still have) a situation a few months back which stumped us all until I could figure out what was happening. During event execution a check failed and the stack was not cleared which over time led to a problem. (The actual bug is fixed, but exactly this issue has been on the back of my mind since.)

Are you referring to the `Locate` inside `Event` vs `Extract` bug that affected clothing in the StdLib? (See: #57)

### Controlling Execution Failures

> There is no way the author can protect against some of these checks and limits, e.g. the limits are not accessible programmatically so you would need to keep those as attributes as well (or instead) and do the limit check "manually"/explicitly.

Yes, I've realized this — and how cumbersome a task this can easily become, especially in libraries. E.g. I've been experimenting on extending the StdLib to contemplate "transparent containers", which means that a `container` might be closed but its contents visible from the outside (i.e. `NOT opaque`). This would require an `Extract` clause that prevents extracting if the container `IS closed`, but because of how `Verb`s and `Event`s would currently end abruptly in case of an `Extract` failure, this means that before attempting to dislocate an object we must always check if it's inside a close container first — this adds a huge burden not only on the library code, but also (and especially) on end users who might wish to extend the library with custom verbs, for it adds a huge layer of complexity to user code.

> I have been thinking about ways to given the author control over, at least some of, these "interruptions". One thought would be to introduce a "try/catch"-like construct where you could catch the "exception" that a failed check might be viewed as.
> 
> Another option could be to allow "dry-running" a statement before actually executing it. \[…\]
> 
> A variation on this might be to allow just trying or accessing the check/limit explicitly.

All three solutions seem interesting (and not necessarily mutually exclusive). I wonder which one is easier implementation-wise.

The "dry run" approach seems the easier one, especially for non programmers. The "try/catch"-like solution might involve some unnecessary complexity (i.e. details about what failed where), whereas in most cases knowing if a certain course of action is viable or not is more than enough (no need to know where exactly it failed).

I like the syntax of the example you provided:

```alan
If Locate obj in hero Will succeed Then ...
```

I find it semantically clean and intuitive.

Trying or accessing the check/limit explicitly might be useful in some other contexts, for decision making, but probably in this scenario it would add unnecessary verbosity compared to the above `If ... will succeed Then/Else` construct. For example, making checks and limits accessible to the adventure could be useful in places where a random object/actor needs to be picked at the current location for some event, rule or script — but that's another story altogether.

Right now, being able to get a boolean evaluation on whether some instructions could potentially end the verb execution seems the most viable solution.

But it would be important to be aware of all the possible instructions that could break up the verb execution on failure. From what I remember, the _ALAN Manual_ only mentions this in containers `Limits` and actors `Script`s. Maybe also `Event`s could end abruptly if they carry out some failing instructions? (we saw it with the clothing bug already) 

### New Features Implementation

> Any either one would be complicated to introduce in the language and in the system.

Which one would the easiest one to implement? And how complex would it be to add anyone of these? Are there any language disruption risks involved?

I see that lack of coder control on these is a serious limitation, as it's also the inability to interrupt manually verbs execution (e.g. during a `DOES BEFORE/AFTER`, to end the execution inheritance chain). This is an area of great interest when it comes to controlling adventures — Verbs are, after all, the core of an IF game, for they interface the player intentions with the Hero's actions in the adventure world; so authors do need to be in control over these executions chains.

PS: I'm still studying ARun sources to get a clearer picture of how the AMachine works, and I'm starting to get a better picture now. I hope that I might be able to write something on AMachine internals soon, for its specs would be really helpful to understand the nature of these problems, as well as the directions of any solution.

### Discussing Lang Limits in Library Contexts

> (As a side note, it would be interesting to talk about things that the libraries tries to do that the Alan language does not provide, perhaps there are valuable language features to be found there.)

Definitely, we should. Right now I'm fresh on such issues, but in six months time I might forget most of them, so it's worth jotting down notes somewhere — even if these are not going to be used immediately.

If it's OK with you, I'd like to use Issues on the Alan repo on GitHub, using some dedicate label (e.g. `lang limits`) to allow quick filtering. There are plenty of library related lang limits to discuss, and it might be best to deal with each one individually, one issue for each problem. I prefer GitHub Issues because they are more familiar to me, can be cross-referenced anywhere on GitHub, and are easier to navigate than the Alan-IF Yahoo Group thanks to labels, milestones and projects.

**Comment by thoni56** [comment]
I've pondering this a bit in sub-concious mode, and I start to lean towards a try/catch solution. I have not even looked at the source code to figure out if this is the most feasible solution or not. It just "feels" better. It does not fundamentally change the language, as the "If <statement> Will Succeed" does. But again, no promises or commitments here at al.

The idea of a `lang limits` is good. The label could probably have a better name, but I can't come up with one now, and it's easy to change when and if we find one.

**Comment by tajmone** [comment]
> I start to lean towards a try/catch solution.

Sound promising (except for the added `setjump()`/`longjump()`s that will be added to the C sources, which make it hard to follow 😉).

> But again, no promises or commitments here at al.

Sure, but seeding the idea is always the first step, then a natural time in which to implement it will present itself (a major refactor, or the like).

> The label could probably have a better name

Maybe `limitations`? anyhow, labels can be easily renamed at any time.

**Comment by tajmone** [comment]
@thoni56: just buzzing you to bring your attention on this Issue again. Any news on this front of providing Try/Catch or other type of dry-runs to enhance verbs definitions?

A long time has passed, so I too had forgotten about it and needed to read through the thread again.
